Extract Environment

Allcharles opened the pull request:
Extracted the environment variables from the application for the inevitable push to deployment.

This should allow a build environment to easily select an environment to push with the build to switch it between development, staging, and production. Currently this environment file only stores the external configs location on the BAW Server (currently redirects to workbench-client as route does not exist).

With this update there are three types of environments:

- Build environment
  - This is baked into the application and cannot be changed after build
- Deployment environment
  - This is the environment which can be changed after build time. This stores information about whether the system is accessing the staging or production server
- Config
  - This is the downloaded config file from the BAW Server. It stores all values that should change between staging and production versions of the application.

Unit tests are a little skimp as they relate to Issue #97 

Closes #90 
Closes #91

Allcharles committed:
Created EnvironmentBuilder

Allcharles committed:
Initial attempt at extracting environment

Allcharles committed:
Fixed production config downloading

Allcharles committed:
Update staging environment to use https

Allcharles committed:
Modified external environment system

Allcharles committed:
Updated app-config unit tests

Allcharles committed:
Added app version to environment

Allcharles committed:
Added environment sanity check

Allcharles committed:
Updated appConfigMock

atruskie commented on src/app/app.helper.ts:
```diff
@@ -3,20 +3,15 @@ import { APP_INITIALIZER } from "@angular/core";
-  APP_CONFIG as MOCK_APP_CONFIG,
-  MockAppConfigService
-} from "./services/app-config/appConfigMockService";
+import { MockAppConfigService } from "./services/app-config/appConfigMock.service";
```
I guess I'll find out what `MockAppConfig` is below... but the name is suspicious. Mocks should only be used in tests...

atruskie replied:
Yeah, it is a test feature. Code in test files, even if just a named constant, should never bleed back into source code.

Try inverting it:
- Define an APP_CONFIG symbol in app-config and override the injection of that symbol in the tests only.

Allcharles replied:
`APP_CONFIG` is no longer used, i have however taking all of the testing code out of `app.helper.ts` and into `test.helper.ts`.

atruskie commented on src/app/services/app-config/app-config.service.ts:
```diff
@@ -13,53 +14,41 @@ export class AppConfigService {
+      if (data.environment.environment !== environment.environment) {
+        throw new Error(
+          `AppConfigService: Deployment environment (${environment.environment}) and config environment (${data.environment.environment}) do not match.`
+        );
```
I'm not sure in which case this condition would occur. Even if it does, why is it an error?

atruskie commented on src/app/services/app-config/app-config.service.ts:
```diff
@@ -93,11 +82,17 @@ export class AppConfigService {
+      environment.environment = options.environment;
+      return options.configUrl;
+    })
+    .then((configUrl: string) => fetch(configUrl))
```
Ok officially confused. Why are we fetching two things from the server?

atruskie reviewed (changes requested):
Okay.

Delete the following files:

- src/assets/config/development.json
- src/assets/config/production.json
- src/assets/config/staging.json
- src/assets/config/
- src/assets/environment/environment.json
- src/assets/environment/environment.production.json
- src/assets/environment/environment.staging.json

Now add: 

- src/assets/environment.json

The contents of which will have the structure of the  `Configuration` interface.

Then, simplify the rest of this code.

Allcharles committed:
Removed external config

Simplified deployment environment files to one file, combined it with config, and moved to /assets/environment.json.

Allcharles committed:
Merge branch 'master' into extract-environment

Allcharles committed:
Extracted testing code from app.helper.ts

Allcharles committed:
Linted project
